homeassistant/components/sensor/geo_rss_events.py

```diff
@@ -0,0 +1,328 @@
+            for event in self._state:
+                matrix[event.title] = '{:.0f}km'.format(event.distance)
+        return matrix
+
+    def update(self):  # pragma: no cover
+        """Update this sensor from the GeoRSS service."""
+        _LOGGER.debug("About to update sensor %s", self.entity_id)
+        self._data.update()
+        all_events = self._data.events
+        if self._category is None:
+            # Add all events regardless of category.
+            _LOGGER.debug("Adding events to sensor %s: %s", self.entity_id,
+                          all_events)
+            self._state = all_events
+        else:
+            # Group events by category.
+            my_events = []
+            if all_events:
+                for event in all_events:
+                    if event.category == self._category:
+                        my_events.append(event)
+            _LOGGER.debug("Adding events to sensor %s: %s", self.entity_id,
+                          my_events)
+            self._state = my_events
+
+
+class GeoRssServiceData(object):
+    """Provides access to GeoRSS feed and stores the latest data."""
+
+    def __init__(self, hass, home_latitude, home_longitude, url, radius_in_km):
+        """Initialize the update service."""
+        self._hass = hass
+        self._home_coordinates = [home_latitude, home_longitude]
+        self._url = url
+        self._radius_in_km = radius_in_km
+        self.events = None
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):  # pragma: no cover
+        """Retrieve data from GeoRSS feed and store events."""
+        import feedparser
+        feed_data = feedparser.parse(self._url)
+        if not feed_data:
+            _LOGGER.error("Error fetching feed data from %s", self._url)
+        else:
+            events = self.filter_entries(feed_data)
+            self.events = events
+
+    def filter_entries(self, feed_data):
+        """Filter entries by distance from home coordinates."""
+        events = []
+        _LOGGER.info("%s entri(es) available in feed %s",
+                     len(feed_data.entries), self._url)
+        for entry in feed_data.entries:
+            geometry = None
+            if hasattr(entry, 'where'):
+                geometry = entry.where
+            elif hasattr(entry, 'geo_lat') and hasattr(entry, 'geo_long'):
+                coordinates = (float(entry.geo_long), float(entry.geo_lat))
+                geometry = type('obj', (object,),
+                                {'type': 'Point', 'coordinates': coordinates})
+            if geometry:
+                distance = self.calculate_distance_to_geometry(geometry)
+            if distance <= self._radius_in_km:
+                event = self.create_event(entry, distance, geometry)
+                events.append(event)
+        _LOGGER.debug("%s events found nearby", len(events))
+        return events
+
+    @staticmethod
+    def create_event(feature, distance, geometry):
```

Comment:
@pvizeli: True, the sensor's state is just the number of events found in the RSS feed filtered by distance. However, as you can see in the `device_state_attributes` method, I am displaying the title and distance in the info dialogue.
My current approach was to simply extract all information from the RSS feed that appeared useful. In a next iteration of this GeoRSS sensor, I was planning to also amend the UI and for example show a map with pins, or offer the option to order the events found by distance or by date.
Let me have a think about how to reduce the code without losing crucial information.